sys/kern/rman.c

```diff
@@ -1,83 +1,252 @@
+        panic("non-empty region");
+      kfree(M_RES, rg);
+    }
+  }
+  /* TODO: destroy the `rm_lock` after implementing `mtx_desotry`. */
+}
+
+static bool rman_region_reserve_resource(rman_region_t *rg, rman_addr_t start,
+                                         rman_addr_t end, size_t count,
+                                         resource_t *r) {
+  resource_t *s, *t;
+  rman_addr_t rstart, rend, amask;
+  res_flags_t flags = r->r_flags;
+
+  assert(mtx_owned(&rg->rg_rman->rm_lock));
 
-  /* Fits before the first resource on the list ? */
-  resource_t *first_r = TAILQ_FIRST(&rm->rm_resources);
-  rman_addr_t first_start = first_r ? first_r->r_start : rm->rm_end + 1;
-  if (align(start, bound) + count <= first_start) {
-    *start_p = align(start, bound);
+  amask = RMAN_ALIGNMENT_GET(flags);
+  start = RMAN_ALIGN(max(rg->rg_start, start), amask);
+  end = min(rg->rg_end, end);
+  if (end - count + 1 < start)
+    return false;
+
+  /* Does the resource fit before the first resource in the region? */
+  s = TAILQ_FIRST(&rg->rg_resources);
+  if (!s || (s && s->r_start - count >= start)) {
+    r->r_rg = rg;
+    r->r_start = start;
+    r->r_end = start + count - 1;
+    TAILQ_INSERT_HEAD(&rg->rg_resources, r, r_link);
     return true;
   }
 
-  /* Look up first element after which we can place our new resource. */
-  resource_t *curr_r = NULL;
-  TAILQ_FOREACH (curr_r, &rm->rm_resources, r_link) {
-    resource_t *next_r = TAILQ_NEXT(curr_r, r_link);
-    rman_addr_t curr_end = curr_r->r_end + 1;
-    rman_addr_t next_start = next_r ? next_r->r_start : rm->rm_end + 1;
+  /* Look up first element after which we can place our new resoucre. */
+  TAILQ_FOREACH_SAFE (s, &rg->rg_resources, r_link, t) {
+    rman_addr_t gap_start = s->r_end + 1;
+    rman_addr_t gap_end = t ? t->r_start - 1 : rg->rg_end;
 
-    /* Skip elements before `start` address. */
-    if (start >= next_start)
+    /* We surely can't fit the resource here (and `gap_start` overflowed). */
+    if (s->r_end == RMAN_ADDR_MAX)
+      break;
+    /* Skip lower gaps. */
+    if (gap_end - count + 1 < start)
       continue;
-
-    /* Do not consider elements after `end` address. */
-    if (end <= curr_end)
+    /* Stop if we've gone too far. */
+    if (gap_start > end - count + 1)
+      break;
+    /* Too large start address for required alignment. */
+    if (gap_start > RMAN_ADDR_MAX - amask)
       break;
 
-    /* Calculate first & last address of eligible gap. */
-    rman_addr_t first = align(max(curr_end, start), bound);
-    rman_addr_t last = min(next_start - 1, end);
-
-    /* Does region of `count` size fit in the gap at `bound` aligned address? */
-    if (first + count - 1 <= last) {
-      *start_p = align(first, bound);
-      *res_p = curr_r;
+    rstart = RMAN_ALIGN(max(start, gap_start), amask);
+    /* Check for overflow. */
+    if ((rstart + count - 1) < rstart)
+      break;
+    rend = min(gap_end, max(rstart + count - 1, end));
```

Comment:
Evalution of `max(rstart + count)` seems to be unnecessary here. `rend` is only used at line _199_:
```C
if (rstart <= rend && rend - rstart >= count - 1) {
```
and after substituting `rend` with `rstart + count - 1` the condition of the _if statement_ becomes
```C
(rstart + count - 1) - rstart >= count - 1
```
This is obviously **always true**.

I suggest replacing `max(rstart + count - 1, end)` with `end`.

Reply:
Uh, never mind. That was a stupid remark.

However I'm still confused by this line.
I don't understand tho why do you compare `rstart + count - 1` with `end`. Is there really a case when `rstart + count - 1` is greater than `end`? Wouldn't that mean that there's not enough space for the resource?
I think that lines _186-188_ and _162-163_ already handle such scenarios.

Reply:
Resolved in e3b65cc.